cluster-provision/gocli/cmd/provision_manager.go

```diff
@@ -0,0 +1,417 @@
+	for candid != "." && candid != "/" {
+		candid = filepath.Dir(candid)
+		match, ok = rulesDB[candid+"/*"]
+		if ok {
+			return match, nil
+		}
+	}
+
+	if status != FILE_DELETED {
+		return nil, fmt.Errorf("Failed to find a rule for " + fileName)
+	}
+
+	return nil, nil
+}
+
+func getKubevirtciTag() (string, error) {
+	const kubevirtciTagUrl = "https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirtci/latest?ignoreCache=1"
+	output, err := runCommand("curl", []string{kubevirtciTagUrl})
+	if err != nil {
+		return "", err
+	}
+
+	return strings.TrimSuffix(output, "\n"), err
+}
+
+func runCommand(command string, args []string) (string, error) {
+	var stdout, stderr bytes.Buffer
+
+	cmd := exec.Command(command, args...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	if err != nil {
+		if strings.Contains(stderr.String(), "unknown revision or path not in the working tree") {
+			logrus.Error("Tag not found, run git pull --tags")
+		}
+		return "", errors.Wrapf(err, "Failed to run git command: git %s\nStdout:\n%s\nStderr:\n%s",
+			strings.Join(args, " "), cmd.Stdout, cmd.Stderr)
+	}
+
+	return stdout.String(), nil
+}
+
+func buildRulesDBfromFile(rulesFile string, targets []string) (map[string][]string, error) {
+	inFile, err := fileSystem.Open(rulesFile)
+	if err != nil {
+		return nil, err
+	}
+	defer inFile.Close()
+
+	return buildRulesDB(inFile, targets)
+}
+
+func buildRulesDB(input io.Reader, targets []string) (map[string][]string, error) {
+	rulesDB := make(map[string][]string)
+
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+		tokens := strings.Fields(line)
+		if len(tokens) != 3 || tokens[1] != "-" {
+			return nil, fmt.Errorf("Invalid syntax (should be 'directory - target'), rule: " + line)
+		}
+
+		dir := tokens[0]
+		target := tokens[2]
+
+		switch {
+		case target == "all":
+			rulesDB[dir] = targets
+		case target == TARGET_NONE:
+			rulesDB[dir] = []string{TARGET_NONE}
+		case strings.HasPrefix(target, "!"):
+			target = target[1:]
+			if !isTargetValid(target, targets) {
+				return nil, fmt.Errorf("Invalid target, rule: " + line)
+			}
+			rulesDB[dir] = excludeTarget(target, targets)
+		case target == "regex":
+			directories, _ := globDirectories(dir)
+			for _, dir := range directories {
+				target = strings.ReplaceAll(filepath.Base(dir), "k8s-", "")
+				if !isTargetValid(target, targets) {
+					return nil, fmt.Errorf("Invalid target " + target + ", rule: " + line)
+				}
+				rulesDB[dir+"/*"] = []string{target}
+			}
+		case target == "regex_none":
+			directories, _ := globDirectories(dir)
+			for _, dir := range directories {
+				rulesDB[dir+"/*"] = []string{TARGET_NONE}
+			}
+		default:
+			if !isTargetValid(target, targets) {
+				return nil, fmt.Errorf("Invalid target, rule: " + line)
+			}
+			rulesDB[dir] = []string{target}
+		}
+	}
+
+	printRulesDB(rulesDB)
```

Comment:
Also the output would maybe better suited to be a struct that contains the information extracted from the config, but best would be to have everything captured directly in the config file.

Reply:
It is a map, we sort it and print it.

Maybe you mean related to the yaml structure ?
Note that it renders dynamically the config file, it is not the config file raw.
It globs the directories and apply the targets (both are dynamic information).
So even if so, it would need to marshal.

Ouput example for our config file:
If you still think we need to do something please elborate .

```
.gitignore : [none]                          
LICENSE : [none]                             
Makefile : [none]                            
OWNERS : [none]                              
cluster-provision : [1.24 1.25 1.26-centos9] 
cluster-provision/centos8/* : [1.24 1.25]    
cluster-provision/centos9/* : [1.26-centos9] 
cluster-provision/gocli/* : [1.24 1.25 1.26-centos9] 
cluster-provision/images/* : [none]          
cluster-provision/k8s : [1.24 1.25 1.26-centos9] 
cluster-provision/k8s/1.24/* : [1.24]        
cluster-provision/k8s/1.25/* : [1.25]        
cluster-provision/k8s/1.26-centos9/* : [1.26-centos9] 
cluster-provision/tools/* : [none]           
cluster-up : [1.24 1.25 1.26-centos9]        
cluster-up/cluster : [1.24 1.25 1.26-centos9] 
cluster-up/cluster/external : [none]         
cluster-up/cluster/k3d-1.25-sriov/* : [none] 
cluster-up/cluster/k3d/* : [none]            
cluster-up/cluster/k8s-1.24/* : [1.24]       
cluster-up/cluster/k8s-1.25/* : [1.25]       
cluster-up/cluster/k8s-1.26-centos9/* : [1.26-centos9] 
cluster-up/cluster/kind-1.23-sriov/* : [none] 
cluster-up/cluster/kind-1.23/* : [none]      
cluster-up/cluster/kind-1.25-vgpu/* : [none] 
cluster-up/cluster/kind/* : [none]           
cluster-up/cluster/local : [none]            
cluster-up/hack : [1.24 1.25 1.26-centos9]   
hack : [1.24 1.25 1.26-centos9]              
hack/kustomization/* : [1.24 1.25 1.26-centos9] 
hack/pman/* : [1.24 1.25 1.26-centos9]       
publish.sh : [1.24 1.25 1.26-centos9]    
```